src/daft-core/src/array/ops/cast.rs

```diff
@@ -1527,6 +1609,308 @@ impl TensorArray {
     }
 }
 
+macro_rules! implement_cast_for_dense_with_inner_dtype {
+    ($dtype:ty, $array_type:ty, $n_values:ident, $non_zero_indices_array:ident, $non_zero_values_array:ident, $offsets:ident) => {{
+        let mut values = vec![0 as $dtype; $n_values];
+        let validity = $non_zero_values_array.validity();
+        for (i, indices) in $non_zero_indices_array.into_iter().enumerate() {
+            let is_valid = validity.map_or(true, |v| v.get_bit(i));
+            if !is_valid {
+                continue;
+            }
+            for j in 0..indices.unwrap().len() {
+                let index = $non_zero_indices_array
```

Comment:
we should move these 
```
 let index_array = $non_zero_indices_array
                    .get(i)
                    .unwrap()
                    .u64()
                    .unwrap();
let values_array = $non_zero_values_array
                    .get(i)
                    .unwrap()
                    .downcast::<$array_type>()
                    .unwrap()
                    .as_arrow();

```
to the outer loop. and then iterate over the values via
for (idx, val) in index_array.into_iter().zip(values_array.into_iter()) {

}
```

Reply:
This will let us avoid bounds checks and let this be vectorized.